File: backend/src/tripl/api/v1/event_types.py
import uuid

from fastapi import APIRouter

from tripl.api.deps import BranchIdDep, EditorUserDep, SessionDep
from tripl.models.event_type import EventType
from tripl.schemas.event_type import EventTypeCreate, EventTypeResponse, EventTypeUpdate
from tripl.schemas.schema_drift import (
    SchemaDriftActionRequest,
    SchemaDriftListResponse,
    SchemaDriftResponse,
)
from tripl.services import audit_service, event_type_service, schema_drift_service

router = APIRouter(prefix="/projects/{slug}/event-types", tags=["event-types"])


@router.get("", response_model=list[EventTypeResponse])
async def list_event_types(
    session: SessionDep, slug: str, branch_id: BranchIdDep
) -> list[EventTypeResponse]:
    return await event_type_service.list_event_types(session, slug, branch_id)


@router.post("", response_model=EventTypeResponse, status_code=201)
async def create_event_type(
    session: SessionDep,
    slug: str,
    data: EventTypeCreate,
    current_user: EditorUserDep,
    branch_id: BranchIdDep,
) -> EventType:
    et = await event_type_service.create_event_type(session, slug, data, branch_id)
    await audit_service.record(
        session,
        user=current_user,
        action="event_type.create",
        target_type="event_type",
        target_id=et.id,
        target_name=et.name,
        project_slug=slug,
        payload=data.model_dump(),
    )
    return et


@router.get("/{event_type_id}", response_model=EventTypeResponse)
async def get_event_type(
    session: SessionDep,
    slug: str,
    event_type_id: uuid.UUID,
    branch_id: BranchIdDep,
) -> EventType:
    return await event_type_service.get_event_type(session, slug, event_type_id, branch_id)


@router.patch("/{event_type_id}", response_model=EventTypeResponse)
async def update_event_type(
    session: SessionDep,
    slug: str,
    event_type_id: uuid.UUID,
    data: EventTypeUpdate,
    current_user: EditorUserDep,
    branch_id: BranchIdDep,
) -> EventType:
    et = await event_type_service.update_event_type(session, slug, event_type_id, data, branch_id)
    await audit_service.record(
        session,
        user=current_user,
        action="event_type.update",
        target_type="event_type",
        target_id=et.id,
        target_name=et.name,
        project_slug=slug,
        payload=data.model_dump(exclude_unset=True),
    )
    return et


@router.delete("/{event_type_id}", status_code=204)
async def delete_event_type(
    session: SessionDep,
    slug: str,
    event_type_id: uuid.UUID,
    current_user: EditorUserDep,
    branch_id: BranchIdDep,
) -> None:
    existing = await event_type_service.get_event_type(session, slug, event_type_id, branch_id)
    name = existing.name
    await event_type_service.delete_event_type(session, slug, event_type_id, branch_id)
    await audit_service.record(
        session,
        user=current_user,
        action="event_type.delete",
        target_type="event_type",
        target_id=event_type_id,
        target_name=name,
        project_slug=slug,
    )


@router.get("/{event_type_id}/drifts", response_model=SchemaDriftListResponse)
async def list_event_type_drifts(
    session: SessionDep, slug: str, event_type_id: uuid.UUID
) -> SchemaDriftListResponse:
    return await schema_drift_service.list_drifts_for_event_type(session, slug, event_type_id)


@router.post("/drifts/{drift_id}/actions", response_model=SchemaDriftResponse)
async def apply_schema_drift_action(
    session: SessionDep,
    slug: str,
    drift_id: uuid.UUID,
    data: SchemaDriftActionRequest,
    current_user: EditorUserDep,
) -> SchemaDriftResponse:
    drift = await schema_drift_service.apply_drift_action(
        session,
        slug,
        drift_id,
        data,
        current_user,
    )
    await audit_service.record(
        session,
        user=current_user,
        action=f"schema_drift.{data.action}",
        target_type="schema_drift",
        target_id=drift_id,
        target_name=drift.field_name,
        project_slug=slug,
        payload=data.model_dump(),
    )
    return drift

File: frontend/src/api/eventTypes.ts
import { api, withBranch } from './client'
import type { EventType, SchemaDriftList, Sensitivity } from '../types'

export const eventTypesApi = {
  list: (slug: string, branchId?: string | null) =>
    api.get<EventType[]>(withBranch(`/projects/${slug}/event-types`, branchId)),
  get: (slug: string, id: string, branchId?: string | null) =>
    api.get<EventType>(withBranch(`/projects/${slug}/event-types/${id}`, branchId)),
  create: (
    slug: string,
    data: {
      name: string
      display_name: string
      description?: string
      color?: string
      field_definitions?: Array<{
        name: string
        display_name: string
        field_type: string
        is_required?: boolean
        sensitivity?: Sensitivity
        contract_required_max_null_rate?: number | null
        contract_regex?: string | null
        contract_min_value?: number | null
        contract_max_value?: number | null
        contract_max_bad_rate?: number
      }>
    },
    branchId?: string | null,
  ) => api.post<EventType>(withBranch(`/projects/${slug}/event-types`, branchId), data),
  update: (
    slug: string,
    id: string,
    data: Partial<{ display_name: string; description: string; color: string; order: number }>,
    branchId?: string | null,
  ) => api.patch<EventType>(withBranch(`/projects/${slug}/event-types/${id}`, branchId), data),
  del: (slug: string, id: string, branchId?: string | null) =>
    api.del(withBranch(`/projects/${slug}/event-types/${id}`, branchId)),
  listDrifts: (slug: string, eventTypeId: string) =>
    api.get<SchemaDriftList>(`/projects/${slug}/event-types/${eventTypeId}/drifts`),
  applyDriftAction: (
    slug: string,
    driftId: string,
    data: {
      action: 'accept' | 'snooze' | 'false_positive' | 'reopen'
      note?: string | null
      snoozed_until?: string | null
    },
  ) => api.post(`/projects/${slug}/event-types/drifts/${driftId}/actions`, data),
}
